src/trace.zig

```diff
@@ -284,67 +227,73 @@ pub const Tracer = struct {
+        const us_log_threshold_ns = 5 * std.time.ns_per_ms;
 
-        const stack = event.stack();
-        const event_start_ns = tracer.events_started[stack].?;
-        const event_end_ns = tracer.timer.read();
+        const event_tracing = event.as(EventTracing);
+        const event_timing = event.as(EventTiming);
+        const stack = event_tracing.stack();
+
+        const event_start = tracer.events_started[stack].?;
+        const event_end = stdx.Instant.now();
+        const event_duration = event_end.duration_since(event_start);
 
         assert(tracer.events_started[stack] != null);
         tracer.events_started[stack] = null;
 
-        log.debug("{}: {}: stop:{} (duration={}ms)", .{
+        // Double leading space to align with 'start: '.
+        log.debug("{}: {s}({}): stop:  {} (duration={}{s})", .{
             tracer.replica_index,
-            event,
-            struct_format(data, .dense),
-            @divFloor(event_end_ns - event_start_ns, std.time.ns_per_ms),
+            @tagName(event),
+            event_tracing,
+            event_timing,
+            if (event_duration.nanoseconds < us_log_threshold_ns)
+                event_duration.microseconds()
+            else
+                event_duration.milliseconds(),
+            if (event_duration.nanoseconds < us_log_threshold_ns) "us" else "ms",
         });
 
-        tracer.write_stop(stack, data);
+        tracer.timing(event_timing, event_duration.microseconds());
+
+        tracer.write_stop(stack);
     }
 
-    pub fn reset(tracer: *Tracer, event_tag: Event.EventTag) void {
-        const stack_base = Event.event_stack_base.get(event_tag);
-        const cardinality = Event.event_stack_cardinality.get(event_tag);
+    pub fn cancel(tracer: *Tracer, event_tag: Event.Tag) void {
+        const stack_base = EventTracing.stack_bases.get(event_tag);
+        const cardinality = EventTracing.stack_limits.get(event_tag);
         for (stack_base..stack_base + cardinality) |stack| {
             if (tracer.events_started[stack]) |_| {
-                log.debug("{}: {s}: reset", .{ tracer.replica_index, @tagName(event_tag) });
+                log.debug("{}: {s}: cancel", .{ tracer.replica_index, @tagName(event_tag) });
 
                 tracer.events_started[stack] = null;
-                tracer.write_stop(@intCast(stack), .{});
+                tracer.write_stop(@intCast(stack));
             }
         }
     }
 
-    fn write_stop(tracer: *Tracer, stack: u32, data: anytype) void {
-        comptime assert(@typeInfo(@TypeOf(data)) == .Struct);
-
+    fn write_stop(tracer: *Tracer, stack: u32) void {
         const writer = tracer.options.writer orelse return;
-        const time_now = std.time.Instant.now() catch unreachable;
-        const time_elapsed_ns = time_now.since(tracer.time_start);
-        const time_elapsed_us = @divFloor(time_elapsed_ns, std.time.ns_per_us);
+        const time_now = stdx.Instant.now();
```

Comment:
This is micro-optimization, but I'd maybe pass `time_elapesd` in as a parameter here, so that in `stop` we don't lookup the current time twice